### PR TITLE
TestExecutionContext.WorkerId

### PR DESCRIPTION
**Pull request by null511**
- Implemented read-only `.WorkerId` property on `TestExecutionContext` to
  enable applications to determine which worker spawned the test. Contains
  `TestWorker.Name` when parallel testing is enabled, otherwise `null`.

### CODE REVIEW

**Comment by null511** [comment]
@CharliePoole Sorry for the additional PR, changed the branch name. Similar implementation as previous PR, but through _TestExecutionContext_. I've tested this against my framework implementing NUnit, and it seems to work as expected. When workers = 1 (or 0), the `.WorkerId` property is null; otherwise it contains the `TestWorker.Name`.

Since the property is only assigned immediately before execution, I had to have the get accessor bubble up through _priorContext_.

File: src/NUnitFramework/framework/Internal/TestExecutionContext.cs [hunk 1]
```diff
@@ -360,6 +365,19 @@ internal IWorkItemDispatcher Dispatcher
...
+        /// For builds with out the parallel feature, it is null.
+        /// </summary>
+        public string WorkerId {
+            get {
+                if (_workerId != null) return _workerId;
+                if (_priorContext != null) return _priorContext.WorkerId;
+                return null;
+            }
+            internal set {_workerId = value;}
+        }
```
Comment (CharliePoole) [inline src/NUnitFramework/framework/Internal/TestExecutionContext.cs]: This surprised me. I figured you would set the WorkerId in CompositeWorkItem.CreateChildWorkItems. This is good but to reduce confusion, please add a comment explaining that some items are executed directly by their parents under the same agent.

**Comment by CharliePoole** [comment]
You haven't exposed this to the user in TextContext. While it's possible for users to reference the internal TestExecutionContext class, we really don't want them to do that! Otherwise, it looks good.

**Comment by null511** [comment]
> You haven't exposed this to the user in TextContext. While it's possible for users to reference the internal TestExecutionContext class, we really don't want them to do that! Otherwise, it looks good.

Ah, sorry. I'm still familiarizing myself with the internals. That makes sense, I'll look into this.

File: src/NUnitFramework/framework/Internal/TestExecutionContext.cs (hunk 1 above)
Comment (null511) [inline src/NUnitFramework/framework/Internal/TestExecutionContext.cs]: I've got to admit, this took a lot more thought then expected to get it passed through at the correct time, but it appears to be working properly.

**Comment by null511** [comment]
Unit test added. If your happy with the changes, everything seems to be working fine on my end.

**Comment by CharliePoole** [comment]
Looks ready to merge. @rprouse Will you second that?

**Comment by CharliePoole** [comment]
@null511 Can you pull the latest changes from master?

**Comment by null511** [comment]
Merged. If you'd prefer a rebase, let me know and i can do so in the morning.

**Comment by CharliePoole** [comment]
This is  fine. We just need a second committer to OK the merge.

**Comment by rprouse** [comment]
Looks good. I will merge.

**Pull request merged**